drivers/serial/uart_nrfx_uarte.c

```diff
@@ -623,189 +1132,186 @@ static int uarte_instance_init(struct device *dev,
 			    &uart_nrfx_uarte_driver_api)
 
-#define UARTE_NRF_RTS_CTS_SET(idx)				\
-	.pselcts = DT_NORDIC_NRF_UARTE_UART_##idx##_CTS_PIN,	\
-	.pselrts = DT_NORDIC_NRF_UARTE_UART_##idx##_RTS_PIN
-
-#define UARTE_NRF_RTS_CTS_NOT_SET				\
-	.pselcts = NRF_UARTE_PSEL_DISCONNECTED,			\
-	.pselrts = NRF_UARTE_PSEL_DISCONNECTED
-
-#define UARTE_NRF_IRQ_ENABLED(idx)				\
-	IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),	\
-		DT_NORDIC_NRF_UARTE_UART_##idx##_IRQ_PRIORITY,	\
-		uarte_nrfx_isr,					\
-		DEVICE_GET(uart_nrfx_uarte##idx),		\
-		0);						\
-	irq_enable(DT_NORDIC_NRF_UARTE_UART_##idx##_IRQ)
-
-#define UARTE_TX_BUFFER_SIZE(idx)				\
-	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE <		\
-		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE) ?	\
-	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE :		\
-		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE)
+#define UARTE_CONFIG(idx)						       \
+	.uart_config = {						       \
+		.baudrate = DT_NORDIC_NRF_UARTE_UART_##idx##_CURRENT_SPEED,    \
+		.data_bits = UART_CFG_DATA_BITS_8,			       \
+		.stop_bits = UART_CFG_STOP_BITS_1,			       \
+		.parity = IS_ENABLED(CONFIG_UART_##idx##_NRF_PARITY_BIT)       \
+			  ? UART_CFG_PARITY_EVEN			       \
+			  : UART_CFG_PARITY_NONE,			       \
+		.flow_ctrl = IS_ENABLED(CONFIG_UART_##idx##_NRF_FLOW_CONTROL)  \
+			     ? UART_CFG_FLOW_CTRL_RTS_CTS		       \
+			     : UART_CFG_FLOW_CTRL_NONE,			       \
+	}
 
-#define UARTE_DATA_INT(idx)				\
-	.tx_buffer = uarte##idx##_tx_buffer
+#define UARTE_NRF_RTS_CTS_PINS(idx)					       \
+	.pselcts = COND_CODE_1(IS_ENABLED(UARTE_##idx##_CONFIG_RTS_CTS),       \
+			       (DT_NORDIC_NRF_UARTE_UART_##idx##_CTS_PIN),     \
+			       (NRF_UARTE_PSEL_DISCONNECTED)),		       \
+	.pselrts = COND_CODE_1(IS_ENABLED(UARTE_##idx##_CONFIG_RTS_CTS),       \
+			       (DT_NORDIC_NRF_UARTE_UART_##idx##_RTS_PIN),     \
+			       (NRF_UARTE_PSEL_DISCONNECTED))
+
+#define UARTE_TX_BUFFER_SIZE(idx)					       \
+	(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE <			       \
+		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE) ?		       \
+			CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE :	       \
+			BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))
+
+#define UARTE_NRF_HW_ASYNC(idx) \
+	.timer = NRFX_TIMER_INSTANCE(CONFIG_UART_##idx##_NRF_HW_ASYNC_TIMER)
+
+#define UARTE_ASYNC(idx)						       \
+	static void uarte##idx##_rx_timeout(struct k_work *work)	       \
+	{								       \
+		rx_timeout(DEVICE_GET(uart_nrfx_uarte##idx));		       \
+	}								       \
+	static void uarte##idx##_tx_timeout(struct k_work *work)	       \
+	{								       \
+		tx_timeout(DEVICE_GET(uart_nrfx_uarte##idx));		       \
+	}								       \
+	struct uarte_async_cb uarte##idx##_async = {			       \
+		UARTE_##idx##_NRF_HW_ASYNC,				       \
+		.rx_timeout_handler = uarte##idx##_rx_timeout,		       \
+		.tx_timeout_handler = uarte##idx##_tx_timeout,		       \
+	}
 
-#define UARTE_CONFIG_INT(idx)				\
-	.tx_buff_size = UARTE_TX_BUFFER_SIZE(idx)
+#define UARTE_INT_DRIVEN(idx)						       \
```

Comment:
What about using instead something like this:
```
#define UARTE_INT_DRIVEN(idx)						       \
	COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),	       \
		(static u8_t uarte##idx##_tx_buffer[			       \
			max(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE,	       \
			    BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))];      \
		 static struct uarte_nrfx_int_driven			       \
		 	uarte##idx##_int_driven = {			       \
				.tx_buffer = uarte##idx##_tx_buffer,	       \
				.tx_buff_size = sizeof(uarte##idx##_tx_buffer),\
			};),						       \
		())
```
Then you could add just: `UARTE_INT_DRIVEN(idx)` to `UART_NRF_UARTE_DEVICE(idx)` and initialize `uarte_##idx##_data` this way:
```
	static struct uarte_nrfx_data uarte_##idx##_data = {		       \
		UARTE_CONFIG(idx),					       \
		COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),  \
			(.int_driven = &uarte##idx##_int_driven),	       \
			()),						       \
	}
```
Same could be done for `ASYNC` stuff and you'll no longer need to use `#ifdef CONFIG_UART_x_ASYNC` and `#ifdef CONFIG_UART_x_INTERRUPT_DRIVEN` for every instance.
Additionally, the fields `int_driven` and `async` won't need to be always present in `uarte_nrfx_data`.